Hermes/appscale/hermes/__init__.py

```diff
@@ -195,67 +164,93 @@ def create_xmpp_user(password, uaserver):
   except UAException as error:
     logging.error('Error while creating an XMPP user: {}'.format(error))
     return False
 
+
 def signal_handler(signal, frame):
   """ Signal handler for graceful shutdown. """
   logging.warning("Caught signal: {0}".format(signal))
   IOLoop.instance().add_callback(shutdown)
 
+
 def shutdown():
   """ Shuts down the server. """
   logging.warning("Hermes is shutting down.")
   IOLoop.instance().stop()
 
+
 def main():
   """ Main. """
-
-  logging_level = logging.INFO
-  if hermes_constants.DEBUG:
-    logging_level = logging.DEBUG
-  logging.getLogger().setLevel(logging_level)
+  parser = argparse.ArgumentParser()
+  parser.add_argument('-p', '--port', type=int, default=DEFAULT_HERMES_PORT,
+                      help='The port to listen on')
+  parser.add_argument('-v', '--verbose', action='store_true',
+                      help='Output debug-level logging')
+  parser.add_argument('--write-profile-log', action='store_true',
+                      help='Write CSV tables with stats on master node')
+  parser.add_argument('--track-processes-stats', action='store_true',
+                      help='Track resources used by monit processes')
+  parser.add_argument('--verbose-cluster-stats', action='store_true',
+                      help='Collect all available stats on master. By default '
+                           'master node requests only the most important stats '
+                           'fields.')
+  args = parser.parse_args()
+
+  logging.basicConfig(format=LOG_FORMAT, level=logging.INFO)
+  if args.verbose:
+    logging.getLogger().setLevel(logging.DEBUG)
+
+  options.define('secret', appscale_info.get_secret())
+  options.define('port', args.port)
 
   signal.signal(signal.SIGTERM, signal_handler)
   signal.signal(signal.SIGINT, signal_handler)
 
-  parse_command_line()
+  master = appscale_info.get_private_ip() == appscale_info.get_headnode_ip()
 
-  app = tornado.web.Application([
-    (MainHandler.PATH, MainHandler),
-    (TaskHandler.PATH, TaskHandler),
-  ], debug=False)
+  if master:
+    # Periodically checks if the deployment is registered and uploads the
+    # appscalesensor app for registered deployments.
+    PeriodicCallback(deploy_sensor_app,
+                     constants.UPLOAD_SENSOR_INTERVAL).start()
 
-  try:
-    app.listen(options.port)
-  except socket.error:
-    logging.error("ERROR on Hermes initialization: Port {0} already in use.".
-      format(options.port))
-    shutdown()
-    return
+    # Periodically check with the portal for new tasks.
+    # Note: Currently, any active handlers from the tornado app will block
+    # polling until they complete.
+    PeriodicCallback(poll, constants.POLLING_INTERVAL).start()
 
-  logging.info("Hermes is up and listening on port: {0}.".
-    format(options.port))
+    # Only master Hermes node handles /do_task route
+    task_route = ('/do_task', TaskHandler)
+  else:
+    task_route = ('/do_task', Respond404Handler,
```

Comment:
I think Hermes master is more like different application from Hermes slave. I don't think it is handler's responsibility to explain why this handler is not supported on slave node.
And `404 Hermes slaves do not manage tasks from Portal` seems representative enough to me.